feat(widget): add circle widget

fujiapple852 opened the pull request:
Adds a simple `Circle` widget.  

This was something I [needed](https://github.com/fujiapple852/trippy/issues/505) to show an accuracy radius on the world map and thought it may be useful to upstream.  There may be more efficient circle drawing algorithms, especially for low-res terminal UIs, this is just the naive textbook implementation.

joshka commented:
Good idea.

Naive is probably fine, though if you care about it you can avoid 7/8s of the trig calls by reflecting at 45 degrees and plotting that point 8 times. (e.g. at (x,y),(y,x),(-y,x),(-x,y),(-x,-y),(-y,-x),(x,-y),(y,-x)). And Bresenham's algorithm is there if you want ever need to avoid the calls altogether and also decrease (or increase for large circles) the number of points actually plotted.

One thing missing on this is tests. Would you mind adding some? Unit tests can be as simple as:
```rust
let mut buffer = Buffer::empty(Rect::new(0, 0, 10, 10));
let canvas = ...

canvas.render(buffer.area, buffer);

let expected = Buffer::with_lines(vec![
  "    ....     "
  "  .      .   "
]);
assert_eq(buffer, expected);
```
Also, perhaps add a circle to the demo / canvas examples?

fujiapple852 commented:
@joshka I've added a unit test (trickier than you'd think to create a unit-test sized example!) and also added a Circle to the the demo app on the map.

fujiapple852 commented:
I [asked](https://www.phind.com/search?cache=03eb5469-d3f4-4aa0-96a8-4a283d4f6373) phind.com (basically ChatGPT 4) to apply your PR comment feedback to the algorithm and it come up with the following:

```rust
impl Shape for Circle {
    fn draw(&self, painter: &mut Painter<'_, '_>) {
        let mut x = 0;
        let mut y = self.radius as i64;
        let mut d = 3 - 2 * (self.radius as i64);

        while x <= y {
            let points = [
                (self.x + x as f64, self.y + y as f64),
                (self.x + y as f64, self.y + x as f64),
                (self.x - x as f64, self.y + y as f64),
                (self.x - y as f64, self.y + x as f64),
                (self.x + x as f64, self.y - y as f64),
                (self.x + y as f64, self.y - x as f64),
                (self.x - x as f64, self.y - y as f64),
                (self.x - y as f64, self.y - x as f64),
            ];

            for &(x, y) in points.iter() {
                if let Some((x, y)) = painter.get_point(x, y) {
                    painter.paint(x, y, self.color);
                }
            }

            if d < 0 {
                d += 4 * x as i64 + 6;
            } else {
                d += 4 * (x as i64 - y as i64) + 10;
                y -= 1;
            }
            x += 1;
        }
    }
}

```

I tried it locally and it works, though I thought the circles weren't as round for low res terminal UIs.  So I'd say better to stick to the naive algorithm.

joshka commented on src/widgets/canvas/circle.rs:
```diff
@@ -0,0 +1,62 @@
+            .x_bounds([-10.0, 10.0])
+            .y_bounds([-10.0, 10.0]);
+        canvas.render(buffer.area, &mut buffer);
+        let expected = Buffer::with_lines(vec![
+            "     ⢀⣠⢤⣀ ",
+            "    ⢰⠋  ⠈⣇",
+            "    ⠘⣆⡀ ⣠⠇",
+            "      ⠉⠉⠁ ",
+            "          ",
+        ]);
+        assert_eq!(buffer, expected);
```
I've added an assert_buffer_eq in #161 that makes tests like this quite a bit easier to write (not merged yet, but it would let you just `let expected = Buffer::with_lines(vec!["          ";5]);` then copy paste the actual result. Saves having to deal with the buffer debug output craziness. It's not merged yet, take a look though and see if you think it would help unit tests generally.

fujiapple852 replied:
Yes that would be useful; I wrote a quick'n'dirty one-liner to collect the symbols to a string

joshka reviewed (approved):
LGTM - the make it fast part of this is not blocking

fujiapple852 commented on examples/demo/ui.rs:
```diff
@@ -346,6 +346,12 @@ where
                 height: 10.0,
                 color: Color::Yellow,
             });
+            ctx.draw(&Circle {
+                x: app.servers[2].coords.1,
+                y: app.servers[2].coords.0,
+                radius: 10.0,
+                color: Color::Green,
+            });
```
That's odd, #works-on-my-computer(tm):

<img width="985" alt="Screenshot 2023-05-02 at 8 15 30 AM" src="https://user-images.githubusercontent.com/36075719/235553408-fe926eec-7c3a-4aef-b942-cf1896fbd57a.png">

What OS/shell are you using?  I'm using macOS/iTerm2.  Your output looks significantly less detailed.

joshka replied:
Weird - tried it again just now and it showed up fine.

joshka replied:
Oh and Mac/iTerm2 also
